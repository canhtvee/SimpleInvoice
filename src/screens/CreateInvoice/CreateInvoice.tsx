import React from 'react';
import {Text, View} from 'react-native';

export function CreateInvoice() {
  return (
    <View>
      <Text>CreateInvoice</Text>
    </View>
  );
}
