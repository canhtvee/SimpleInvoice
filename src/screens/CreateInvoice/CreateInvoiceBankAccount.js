import React from 'react';
import {useFormContext} from 'react-hook-form';

import {AppText, AppTextInput} from '../../commons';
import {Sizes} from '../../utils';
import {View} from 'react-native';

export function CreateInvoiceBankAccount() {
  const {control} = useFormContext();

  return (
    <>
      <AppText style={{marginTop: Sizes.space3x, fontWeight: 'bold'}}>
        Bank Account Info
      </AppText>
      <AppTextInput
        label="bankId"
        control={control}
        name="bankAccount.bankId"
        placeholder={'Enter bankId'}
        containerStyle={{marginTop: Sizes.padding}}
      />
      <AppTextInput
        label="sortCode"
        control={control}
        name="bankAccount.sortCode"
        placeholder={'Enter sortCode'}
        containerStyle={{marginTop: Sizes.padding}}
      />
      <AppTextInput
        label="accountNumber"
        control={control}
        name="bankAccount.accountNumber"
        placeholder={'Enter accountNumber'}
        containerStyle={{marginTop: Sizes.padding}}
      />
      <AppTextInput
        label="accountName"
        control={control}
        name="bankAccount.accountName"
        placeholder={'Enter accountName'}
        containerStyle={{marginTop: Sizes.padding}}
      />
    </>
  );
}
