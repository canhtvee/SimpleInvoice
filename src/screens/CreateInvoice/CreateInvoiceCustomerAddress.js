import React from 'react';
import {useFieldArray, useFormContext} from 'react-hook-form';

import {AppText, AppTextInput} from '../../commons';
import {Sizes} from '../../utils';

export function CreateInvoiceCustomerAddress() {
  const {control} = useFormContext();

  const {fields, append, remove} = useFieldArray({
    control,
    name: 'customer.addresses',
  });

  console.log(fields);

  return (
    <>
      <AppText style={{marginTop: Sizes.space3x, fontWeight: 'bold'}}>
        Customer Info Address
      </AppText>
    </>
  );
}

// customer: {
//   firstName: 'Nguyen',
//   lastName: 'Dung 2',
//   contact: {
//     email: 'nguyendung2@101digital.io',
//     mobileNumber: '+6597594971',
//   },
//   addresses: [
//     {
//       premise: 'CT11',
//       countryCode: 'VN',
//       postcode: '1000',
//       county: 'hoangmai',
//       city: 'hanoi',
//     },
//   ],
// },

const _ = [
  {
    city: 'hanoi',
    countryCode: 'VN',
    county: 'hoangmai',
    id: '10ce8290-dd7f-4148-8e2c-e574727abfa4',
    postcode: '1000',
    premise: 'CT11',
  },
];
