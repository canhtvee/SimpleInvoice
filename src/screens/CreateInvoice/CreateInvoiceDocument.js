import React from 'react';
import {View} from 'react-native';
import {useFieldArray, useFormContext} from 'react-hook-form';

import {AppButtonNormal, AppText, renderFields} from '../../commons';
import {Colors, Sizes, uuid} from '../../utils';

export function CreateInvoiceDocument() {
  const {control} = useFormContext();

  const {fields, append, remove} = useFieldArray({
    control,
    name: 'documents',
  });

  return (
    <>
      <AppText
        style={{
          marginTop: Sizes.space3x,
          fontWeight: 'bold',
          borderRadius: Sizes.borderRadius,
        }}>
        Documents
      </AppText>

      {fields.map((field, index) => {
        const path = `documents.${index}`;

        return (
          <View
            key={field.id}
            style={{
              backgroundColor: Colors.ripple,
              borderRadius: Sizes.borderRadius,
              marginTop: Sizes.padding,
            }}>
            <AppButtonNormal
              title="Delete"
              style={{alignSelf: 'flex-end', marginTop: Sizes.padding}}
              onPress={() => remove(index)}
            />
            {renderFields(
              [
                {key: 'documentId'},
                {key: 'documentName'},
                {key: 'documentUrl'},
              ],
              path,
              {control},
            )}
          </View>
        );
      })}

      <AppButtonNormal
        title="Upload Document"
        style={{alignSelf: 'flex-end', marginTop: Sizes.space3x}}
        onPress={() => {
          const id = uuid();
          append({
            documentId: id,
            documentName: 'Canh',
            documentUrl: `http://url.com/#${id}`,
          });
        }}
      />
    </>
  );
}
