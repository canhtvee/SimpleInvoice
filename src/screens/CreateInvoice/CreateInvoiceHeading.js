import React from 'react';
import {AppText} from '../../commons';
import {Sizes} from '../../utils';

export function CreateInvoiceHeading() {
  return (
    <AppText
      style={{
        alignSelf: 'center',
        fontSize: Sizes.headline,
        fontWeight: 'bold',
        marginTop: Sizes.wpx(60),
      }}>
      Create Invoice
    </AppText>
  );
}
