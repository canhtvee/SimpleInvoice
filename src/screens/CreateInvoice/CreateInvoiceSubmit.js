import React from 'react';
import {Alert} from 'react-native';

import {useNavigation} from '@react-navigation/native';
import {useFormState} from 'react-hook-form';
import uuid from 'react-native-uuid';

import {AppButton} from '../../commons';
import {FetchApi, Sizes} from '../../utils';

import dayjs from 'dayjs';
import {InvoiceService} from '../InvoiceList';
import {mockInvoice} from './mocking';

export function CreateInvoiceSubmit({control, handleSubmit}) {
  const navigation = useNavigation();
  const {isSubmitting} = useFormState({control});

  const onSubmit = async data => {
    const item = {
      itemReference: uuid.v4(),
      description: 'Honda RC' + Math.floor(Math.random() * 150 + 1),
      quantity: Math.floor(Math.random() * 10 + 1),
      amount: Math.floor(Math.random() * 10000 + 1000),
    };

    const baseInvoice = JSON.parse(JSON.stringify(mockInvoice));
    const newInvoice = {
      ...baseInvoice,
      ...data,
      dueDate: dayjs(data.dueDate).format('YYYY-MM-DD'),
    };
    newInvoice.items[0] = {...newInvoice.items[0], ...item};

    console.log(newInvoice);

    try {
      const result = await FetchApi.createInvoice(newInvoice);

      if (!result?.paging) {
        Alert.alert(null, 'Create Invoice Failed');
        throw new Error();
      }

      Alert.alert(null, 'Create Invoice Successful');
      InvoiceService.set(newInvoice);
      navigation.goBack();
    } catch (error) {
      console.log('error', error);
    }
  };

  return (
    <AppButton
      title="Create Invoice"
      onPress={handleSubmit(onSubmit)}
      isLoading={isSubmitting}
      style={{marginVertical: Sizes.wpx(40)}}
    />
  );
}
