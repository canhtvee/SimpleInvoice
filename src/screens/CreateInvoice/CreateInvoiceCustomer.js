import React from 'react';
import {useFormContext} from 'react-hook-form';

import {AppText, renderFields} from '../../commons';
import {Sizes} from '../../utils';
import {CreateInvoiceCustomerAddress} from './CreateInvoiceCustomerAddress';

export function CreateInvoiceCustomer() {
  const {control} = useFormContext();

  return (
    <>
      <AppText style={{marginTop: Sizes.space3x, fontWeight: 'bold'}}>
        Customer Info
      </AppText>
      {renderFields(
        [
          {
            key: 'customer',
            children: [
              {key: 'firstName'},
              {key: 'lastName'},
              {
                key: 'contact',
                children: [{key: 'email'}, {key: 'mobileNumber'}],
              },
            ],
          },
        ],
        '',
        {control, containerStyle: {marginTop: Sizes.padding}},
      )}
      <CreateInvoiceCustomerAddress />
    </>
  );
}
