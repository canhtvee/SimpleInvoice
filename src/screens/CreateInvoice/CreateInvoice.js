import React from 'react';
import {useForm} from 'react-hook-form';
import {KeyboardAwareScrollView} from 'react-native-keyboard-aware-scroll-view';

import {AppContainer, AppDateInput, AppTextInput} from '../../commons';
import {Sizes} from '../../utils';

import {CreateInvoiceHeading} from './CreateInvoiceHeading';
import {CreateInvoiceSubmit} from './CreateInvoiceSubmit';

export function CreateInvoice() {
  const {control, handleSubmit} = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    defaultValues: {
      invoiceReference: '',
      amount: '',
      dueDate: '',
      description: '',
    },
  });

  return (
    <AppContainer>
      <KeyboardAwareScrollView
        contentContainerStyle={{paddingHorizontal: Sizes.wpx(30)}}>
        <CreateInvoiceHeading />

        <AppTextInput
          label="invoiceReference"
          control={control}
          name={'invoiceReference'}
          placeholder={'invoiceReference'}
        />
        <AppTextInput
          label="amount"
          control={control}
          name={'amount'}
          placeholder={'amount'}
        />
        <AppDateInput label="dueDate" control={control} name={'dueDate'} />
        <AppTextInput
          label="description"
          control={control}
          name={'description'}
          placeholder={'description'}
        />

        <CreateInvoiceSubmit handleSubmit={handleSubmit} control={control} />
      </KeyboardAwareScrollView>
    </AppContainer>
  );
}
