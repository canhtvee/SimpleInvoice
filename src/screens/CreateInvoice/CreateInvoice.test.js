import React from 'react';
import {cleanup, render} from '@testing-library/react-native';

import {CreateInvoice} from './CreateInvoice';
import {renderWithNavigator} from '../../utils';

/**
 * Run test command
 * yarn test -u -t="CreateInvoice"
 */

afterEach(cleanup);

jest.mock('react-native-keyboard-aware-scroll-view', () => {
  const {ScrollView} = require('react-native');

  return {
    KeyboardAwareScrollView: ScrollView,
  };
});

describe('CreateInvoice', () => {
  test('should render correctly', () => {
    const tree = renderWithNavigator(<CreateInvoice />);

    expect(tree).toMatchSnapshot();
  });
});
