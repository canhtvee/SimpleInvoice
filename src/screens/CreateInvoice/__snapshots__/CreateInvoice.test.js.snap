// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateInvoice should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "left",
      "right",
    ]
  }
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "flex-start",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingHorizontal": 59,
      }
    }
  >
    <View>
      <Text
        style={
          [
            {
              "color": "#000000",
              "fontSize": 27.5,
            },
            {
              "alignSelf": "center",
              "fontSize": 43.5,
              "fontWeight": "bold",
              "marginTop": 118.5,
            },
          ]
        }
      >
        Create Invoice
      </Text>
      <View>
        <Text
          style={
            [
              {
                "color": "#000000",
                "fontSize": 27.5,
              },
              undefined,
            ]
          }
        >
          invoiceReference
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "#0000008A",
                "borderRadius": 6,
                "borderWidth": 1,
                "flexDirection": "row",
                "overflow": "hidden",
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="#123456"
            placeholderTextColor="#00000061"
            spellCheck={false}
            style={
              [
                {
                  "flex": 1,
                  "fontSize": 27.5,
                  "paddingHorizontal": 16,
                },
                {
                  "paddingHorizontal": 16,
                  "paddingVertical": 23.5,
                },
                {
                  "color": "#000000",
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
      </View>
      <View>
        <Text
          style={
            [
              {
                "color": "#000000",
                "fontSize": 27.5,
              },
              undefined,
            ]
          }
        >
          amount
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "#0000008A",
                "borderRadius": 6,
                "borderWidth": 1,
                "flexDirection": "row",
                "overflow": "hidden",
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="100000"
            placeholderTextColor="#00000061"
            spellCheck={false}
            style={
              [
                {
                  "flex": 1,
                  "fontSize": 27.5,
                  "paddingHorizontal": 16,
                },
                {
                  "paddingHorizontal": 16,
                  "paddingVertical": 23.5,
                },
                {
                  "color": "#000000",
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
      </View>
      <View>
        <Text
          style={
            [
              {
                "color": "#000000",
                "fontSize": 27.5,
              },
              undefined,
            ]
          }
        >
          dueDate
        </Text>
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "overflow": "hidden",
                },
                {
                  "borderColor": "#0000008A",
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "fontSize": 27.5,
                    "justifyContent": "space-between",
                    "paddingHorizontal": 16,
                    "paddingVertical": 23.5,
                  },
                  {
                    "color": "#00000061",
                  },
                ]
              }
            >
              YYYY-MM-DD
            </Text>
          </View>
        </View>
      </View>
      <View>
        <Text
          style={
            [
              {
                "color": "#000000",
                "fontSize": 27.5,
              },
              undefined,
            ]
          }
        >
          description
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "#0000008A",
                "borderRadius": 6,
                "borderWidth": 1,
                "flexDirection": "row",
                "overflow": "hidden",
              },
              undefined,
            ]
          }
        >
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Invoice is issued to Akila Jayasinghe"
            placeholderTextColor="#00000061"
            spellCheck={false}
            style={
              [
                {
                  "flex": 1,
                  "fontSize": 27.5,
                  "paddingHorizontal": 16,
                },
                {
                  "paddingHorizontal": 16,
                  "paddingVertical": 23.5,
                },
                {
                  "color": "#000000",
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={16}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#2196f3",
            "borderRadius": 6,
            "height": 79,
            "justifyContent": "center",
            "marginVertical": 79,
            "opacity": 1,
            "paddingHorizontal": 47,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 29.5,
                "fontWeight": "600",
              },
              {
                "color": "#FFFFFF",
              },
              undefined,
              undefined,
            ]
          }
        >
          Create Invoice
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
